generators/typescript-msw/npm/src/index.ts

```diff
@@ -0,0 +1,93 @@
+
+export type SafeResponse<T extends StandardResponse = StandardResponse> =
+    Omit<T, 'response'>
+    & { headers?: Record<string, unknown> };
+
+export function toMswResponse(response: SafeResponse, res: ResponseComposition, ctx: RestContext) {
+    return res(
+        ctx.status(response.statusCode === 'other' ? 0 : response.statusCode),
+        ctx.json(response.data)
+        // TODO: support headers
+    );
+}
+
+export class MappedRestHandler<T extends MockedRequest = MockedRequest> extends RestHandler<T> {
+    constructor(private request: AdapterRequestArgs, response: SafeResponse | AsyncResponseResolver<T>) {
+        super(request.method, new URL(request.path.split('?')[0], 'http://localhost').toString(),
+        async (req: T, res, ctx): Promise<ResponseResolverReturnType<MockedResponse<any>>> => {
+            if ('body' in request) {
+                if (!deepEqual(request.body, await req.json())) {
+                    this.shouldSkip = true;
+                    return res();
+                }
+            } else if (req.bodyUsed) {
+                this.shouldSkip = true;
+                return res();
+            }
+            // TODO: request headers aren't checked
+
+            if (typeof response === 'function') return response(req, res, ctx);
+            return toMswResponse(response, res, ctx);
+        });
+    }
+
+    override run(...params: Parameters<RestHandler<T>['run']>): ReturnType<RestHandler<T>['run']> {
+        this.shouldSkip = false;
+        return super.run(...params);
+    }
+
+    override predicate(request: T, parsedResult: Match): boolean {
+        // body isn't checked here because this can't be async, so instead it is checked in the MappedRestHandler above
+        return super.predicate(request, parsedResult) && this.queryStringMatch(request);
+    }
+
+    private queryStringMatch(request: T) {
+        return request.url.searchParams.toString() == (this.request.path.split('?')[1] ?? '');
+    }
+}
+
+export function toMswHandler<
+    TMethod extends HttpMethod,
+    TUrlParams extends {},
+    TRequestParams extends TUrlParams,
+    TRequestBodies extends RequestBodies,
+    TResponses extends StandardResponse,
+    TCallType extends TransformCallType
+>(conversion: RequestConversion<TMethod, TUrlParams, TRequestParams, TRequestBodies, TResponses, TCallType>) {
+    type MyRequest =
+        | (TCallType extends 'no-body' | 'optional'
+            ? { params: TRequestParams, body?: never, mimeType?: never }
+            : never)
+        | (TCallType extends 'body' | 'optional'
+            ? { [K in keyof TRequestBodies]: { params: TRequestParams, body?: TRequestBodies[K], mimeType?: K } }[keyof TRequestBodies]
+            : never);
+
+    return function (request: MyRequest, response: AsyncResponseResolver | SafeResponse<TResponses>) {
```

Comment:
The request at least needs to know the URL and parameters (which is why body/mime type are optional), otherwise it would return that Response to all endpoints (which is easy to do with MSW without this binding).

Reply:
URL, yes. Params, not really

The way that we mock out requests in SQT is that we have a top-level server listener that is setup that is a catchall for **all** requests. This eliminates the need to match request params, leading to easier setup and less mental overhead.

When we need a special case for a specific test suite where we expect a certain response, then we override the existing mock for that test suite.

https://mswjs.io/docs/api/setup-server/use#one-time-override

Reply:
Talked about this and makes sense

Reply:
Sounds like we need a future enhancement to URL generation to support `:param` syntax. This sounds really useful, especially if we want to get to an Express server generator. I'm going to hold that for a future enhancement, though. Thanks for the idea!